### PR TITLE
TDL-17096 updated the write_bookmark() logic

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Updated the bookmark written for base streams. 
    - bookmark on the lesser of the maximum replication key value and the start time of sync execution. 
- Added unittests for the same

# QA steps
- Check that the correct bookmark is written based on min-max evaluation
- was pr-alpha'd to two separate clients, but without feedback if it fixed their data discrepancies. 

# Risks
- the tap will likely take longer with extra api calls because of the new bookmarking. performance testing was done in this card: https://jira.talendforge.org/browse/TDL-17096
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_bookmarks.py [hunk 1]
```diff
@@ -19,6 +19,37 @@ def __init__(self, *args, **kwargs):
...
+
+        Comparisons are based on the class of the bookmark value. Dates will be
+        string compared which works for ISO date-time strings
+        """
+        max_bookmarks = {}
+        for stream, batch in sync_records.items():
+            if stream != 'metafields':
+                upsert_messages = [m for m in batch.get('messages') if m['action'] == 'upsert']
+            else:
+                upsert_messages = [m for m in batch.get('messages') if m['action'] == 'upsert' and m.get('data').get('owner_resource') == 'shop']
```
Comment (savan-chovatiya) [inline tests/test_bookmarks.py:35]: Add a comment here regarding the specific behavior of metafields stream.

File: tests/test_bookmarks_updated.py [hunk 2]
```diff
@@ -95,10 +95,21 @@ def bookmarks_test(self, conn_id, testable_streams):
...
+                else:
+                    first_sync_messages = [record.get('data') for record in first_sync_records.get(stream, {}).get('messages', [])
+                                           if record.get('action') == 'upsert' and record.get('data').get('owner_resource') == 'shop']
+
+                if stream != 'metafields':
+                    second_sync_messages = [record.get('data') for record in second_sync_records.get(stream, {}).get('messages', [])
+                                            if record.get('action') == 'upsert']
+                else:
+                    second_sync_messages = [record.get('data') for record in second_sync_records.get(stream, {}).get('messages', [])
+                                        if record.get('action') == 'upsert' and record.get('data').get('owner_resource') == 'shop']
```
Comment (savan-chovatiya) [inline tests/test_bookmarks_updated.py:111]: Add a comment here regarding the specific behavior of metafields stream.

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tests/test_bookmarks.py (hunk 1 above)
Comment (namrata270998) [reply tests/test_bookmarks.py:39]: added comments

File: tests/test_bookmarks_updated.py (hunk 2 above)
Comment (namrata270998) [reply tests/test_bookmarks_updated.py:115]: added comments

**Review by savan-chovatiya — APPROVED** [review]

File: tap_shopify/streams/base.py [hunk 3]
```diff
@@ -234,6 +240,8 @@ def get_objects(self):
                 self.update_bookmark(since_id, bookmark_key='since_id')
 
             updated_at_min = updated_at_max
+        bookmark = max(min(stop_time, max_bookmark), (stop_time - datetime.timedelta(days=date_window_size))) #pylint: disable=line-too-long
```
Comment (luandy64) [inline tap_shopify/streams/base.py:243]:
```suggestion
        bookmark = max(min(stop_time,
                           max_bookmark),
                       (stop_time - datetime.timedelta(days=date_window_size)))
```
I think this is more readable. The original way makes it hard to tell what we are taking the `max` of.

File: tap_shopify/streams/base.py [hunk 4]
```diff
@@ -204,12 +206,16 @@ def get_objects(self):
                     objects = self.call_api(query_params)
 
                 for obj in objects:
+                    dict_obj = obj.to_dict()
+                    replication_value = strptime_to_utc(dict_obj[self.replication_key])
```
Comment (luandy64) [inline tap_shopify/streams/base.py:210]: If performance is a concern here, `to_dict` may not be the fastest option.

As you can see on the line below this, we can access fields on the object with the dot operator.
```python
    if obj.id < since_id:
```

The Python data model actually gives us a way to dynamically do this via `getattr()`
> Return the value of the named attribute of object. name must be a string. ... getattr(x, 'foobar') is equivalent to x.foobar

Source: [Docs Link](https://docs.python.org/3/library/functions.html#getattr)

The code would then look like
```suggestion
                    replication_value = strptime_to_utc(getattr(obj, self.replication_key))
```
Someone should test the performance a little more thoroughly, but here's what I did to performance test and the results:

```python
>>> from timeit import default_timer as timer
>>> def foo():
>>>   start = timer()
>>>   obj.to_dict()[self.replication_key]
>>>   end = timer()
>>>   return end - start
>>> 
>>> sum([foo() for i in range(100)]) / 100
3.0518800485879184e-06
>>> sum([foo() for i in range(100)]) / 100
3.034400288015604e-06
>>> sum([foo() for i in range(100)]) / 100
3.2068253494799135e-06
>>> 
>>> 
>>> sum([bar() for i in range(100)]) / 100
6.84354454278946e-07
>>> sum([bar() for i in range(100)]) / 100
6.862124428153039e-07
>>> sum([bar() for i in range(100)]) / 100
6.863870657980443e-07
```

Summary
```org
   | trial |                to_dict |               getattr |
   |-------+------------------------+-----------------------|
   |     1 | 3.0518800485879184e-06 |  6.84354454278946e-07 |
   |     2 |  3.034400288015604e-06 | 6.862124428153039e-07 |
   |     3 | 3.2068253494799135e-06 | 6.863870657980443e-07 |
```

File: tap_shopify/streams/base.py [hunk 5]
```diff
@@ -204,12 +206,16 @@ def get_objects(self):
...
                     if obj.id < since_id:
                         # This verifies the api behavior expectation we
                         # have that all results actually honor the
                         # since_id parameter.
                         raise OutOfOrderIdsError("obj.id < since_id: {} < {}".format(
                             obj.id, since_id))
+                    if replication_value > max_bookmark:
+                        max_bookmark = replication_value
```
Comment (luandy64) [inline tap_shopify/streams/base.py:218]: # Opinion

 I think we should move `replication_value`'s initialization to just above line 217 here.

It's not used in lines 211 to 216 and only used in this area to check if we should updated `max_bookmark`

File: tap_shopify/streams/base.py (hunk 3 above)
Comment (namrata270998) [reply tap_shopify/streams/base.py]: Updated the code for better readability

File: tap_shopify/streams/base.py (hunk 4 above)
Comment (namrata270998) [reply tap_shopify/streams/base.py]: Used getattr() instead of to_dict()

File: tap_shopify/streams/base.py (hunk 5 above)
Comment (namrata270998) [reply tap_shopify/streams/base.py:217]: Moved the initialization of `replication_value` above line 217

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## 1.7.1
+  * Update bookmarking logic [#143] (https://github.com/singer-io/tap-shopify/pull/143)
```
Comment (dsprayberry) [inline CHANGELOG.md:4]: Space between the bracket and paren will break the link

**Review by dsprayberry — CHANGES_REQUESTED** [review]

**Review by dsprayberry — APPROVED** [review]

**Pull request merged**